File: kotlin/binary-search/src/main/kotlin/BinarySearch.kt
object BinarySearch {
    fun search(list: List<Int>, item: Int): Int {
        var low = 0
        var high = list.size - 1
        while (low <= high) {
            val mid = (low + high) / 2
            if (list[mid] == item) {
                return mid
            } else if (list[mid] < item) {
                low = mid + 1
            } else {
                high = mid - 1
            }
        }
        throw NoSuchElementException()
    }
}

File: fsharp/binary-search/BinarySearch.fs
module BinarySearch

let find (input: int []) search =
    let rec aux lo hi =
        if hi < lo then
            None
        else
            let mid = (lo + hi) / 2
            let candidate = input[mid]
            if candidate = search then Some mid
            elif search < candidate then aux lo (mid - 1)
            else aux (mid + 1) hi
    aux 0 (input.Length - 1)
